src/vscode-dts/vscode.proposed.debugDataBreakpoints.d.ts

```diff
@@ -0,0 +1,162 @@
+		 */
+		constructor(address: string, bytes?: number, description?: string);
+	}
+
+	export class ExpressionDataBreakpointSource {
+		/** A global expression that is first evaluated when the breakpoint is activated. */
+		expression: string;
+		/** A human-readable label for the data breakpoint source. */
+		description: string;
+
+		/**
+		 * Creates a new expression data breakpoint source.
+		 *
+		 * @param expression A global expression that is first evaluated when the breakpoint is activated.
+		 * @param description A human-readable label for the data breakpoint source.
+		 */
+		constructor(expression: string, description?: string);
+	}
+
+	export class VariableScopedDataBreakpointSource {
+		/** Reference to the variable container that has the variable named `variable`. */
+		variablesReference: number;
+		/** The name of the variable that is used for resolution. */
+		variable: string;
+		/** A human-readable label for the data breakpoint source. */
+		description: string;
+
+		/**
+		 * Creates a new variable scoped data breakpoint source.
+		 *
+		 * @param variablesReference Reference to the variable container that has the variable named `variable`.
+		 * @param variable The name of the variable that is used for resolution.
+		 * @param description A human-readable label for the data breakpoint source.
+		 */
+		constructor(variablesReference: number, variable: string, description?: string);
+	}
+
+	export class FrameScopedDataBreakpointSource {
+		/** Reference to the stack frame to which the expression is scoped. */
+		frameId: number;
+		/** The name of the expression that is used for resolution. */
+		expression: string;
+		/** A human-readable label for the data breakpoint source. */
+		description: string;
+
+		/**
+		 * Creates a new frame scoped data breakpoint source.
+		 *
+		 * @param frameId Reference to the stack frame to which the expression is scoped.
+		 * @param expression The name of the expression that is used for resolution.
+		 * @param description A human-readable label for the data breakpoint source.
+		 */
+		constructor(frameId: number, expression: string, description?: string);
+	}
+
+	/** The source for a data breakpoint. */
+	export type DataBreakpointSource = ResolvedDataBreakpointSource | AddressDataBreakpointSource | ExpressionDataBreakpointSource | VariableScopedDataBreakpointSource | FrameScopedDataBreakpointSource;
+
+	/** Access type for data breakpoints. */
+	export type DataBreakpointAccessType = 'read' | 'write' | 'readWrite';
+
+	/**
+	 * A breakpoint specified by a variable or memory change.
+	 */
+	export class DataBreakpoint extends Breakpoint {
+		/** The source for the data breakpoint. See the different source types on how they are resolved during breakpoint activation. */
+		source: DataBreakpointSource;
+
+		/** The resolution of the data breakpoint based on the given source. This may be undefined if the breakpoint was not yet resolved. */
+		resolution?: DataBreakpointResolution;
```

Comment:
similar to below, I wonder if we should not have this at all, but require people use `resolve(session): Promise<DataBreakpointResolution>` to see this information

Reply:
I fully agree, I just adapted the naming here a bit.